--- a/dlls/wined3d/vk_render_pass.c
+++ b/dlls/wined3d/vk_render_pass.c
@@ -23,6 +23,11 @@
 #include "wined3d_private.h"
 WINE_DEFAULT_DEBUG_CHANNEL(wined3d);
+#ifdef __aarch64__
+/* Encourage render pass merging for tile-based renderers */
+#define RENDER_PASS_MERGING 1
+#define MAX_SUBPASSES_TO_MERGE 4
+#endif
+
 static void wined3d_vk_render_pass_cleanup(struct wined3d_vk_render_pass *render_pass)
 {
